app/components/machine/driver-aliyunecs/component.js

```diff
@@ -553,13 +553,13 @@ export default Ember.Component.extend(Driver, {
         errors.push(this.get('intl').t('machine.driverAliyunecs.errors.nameTooShort'));
       }
       //check '.','-' can not be used as the first letter and the last letter
-      if (!/^[a-zA-Z0-9]([\.-]?[a-zA-Z0-9]+[\.-]?)*[a-zA-Z0-9]$/.test(name)){
+      if (!/^[a-zA-Z]([\.]?[a-zA-Z0-9\-]+[\.]?)*[a-zA-Z0-9]$/.test(name)){
```

Comment:
Is 0-9 actually not allowed by them as the first letter?  It is valid in general.  Actually looking closer this entire check is odd because it allows dots, so the hyphen checking needs to be per-label, not the beginning or end of the whole string.

I would suggest pulling this out into an `app/utils/validate-dns.js` with `validateHostname()` and `validateLabel()` functions and redoing them to be simple string manipulation functions rather than regex, other than checking for the basic allowed characters:

An individual `label`:
  - Must consist of the characters a-z (case-insensitive), 0-9, and hyphen (`str.match(/^[a-z0-9-]+$/i)`)
  - Must be 1-63 characters (str.length)
  - Cannot start with or end with a hyphen. (`str.slice(0,1) === '-'`, `str.slice(-1) === '-'`)
  - Cannot contain two consecutive hyphens. (`str.includes('--')`)
  - (CAN start with, end with, or be entirely made up of numbers)

A `hostname`:
  - Must consist of one or more valid `labels`, separated by a single dot. (`str.split(/\./)` and loop)
  - Cannot start with a dot
  - Cannot contain two consecutive dots (i.e. an empty label)
  - CAN end with a dot (this makes it an explicitly fully qualified domain name, but is not always desirable)
  - The total length of the hostname can be at most 253 characters (255 minus one for null-termination, and one for the trailing dot if not already present)

You can add simple unit tests to verify the behavior, see e.g. `tests/unit/utils/parse-port-test.js`

Reply:
@vincent99 Aliyun doesn't allow hostname starting with a number, I just checked. We can only enforce the common part of different cloud provider's policy, not sure if it's a good idea.

Reply:
Yes, I'd like a general dns label validator (which can be used with other providers and other things) and then if that passes you can add additional aliyun specific checks (like this, or their max-length being 120 chars)